src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.GetModuleFileNameLongPath.cs

```diff
@@ -11,67 +11,57 @@ internal static partial class PInvoke
     {
         public static unsafe string GetModuleFileNameLongPath(HINSTANCE hModule)
         {
-            if (TryGetModuleFileName(hModule, out char[] path))
+            Span<char> buffer = stackalloc char[MAX_PATH];
+            uint pathLength;
+            bool isBufferTooSmall = false;
+            fixed (char* lpFilename = buffer)
+            {
+                pathLength = GetModuleFileName(hModule, lpFilename, (uint)buffer.Length);
+                isBufferTooSmall = (WIN32_ERROR)Marshal.GetLastWin32Error() == WIN32_ERROR.ERROR_INSUFFICIENT_BUFFER;
+            }
+
+            if (pathLength == 0)
             {
-                return new string(path);
+                return string.Empty;
             }
 
-            char[] buffer;
+            if (pathLength < buffer.Length - 1 && !isBufferTooSmall)
+            {
+                return new string(buffer[..(int)pathLength]);
+            }
+
+            char[] lbuffer;
             int bufferSize = 4096;
             // Allocate increasingly larger portions of memory until successful or we hit short.maxvalue.
             for (int i = 1; bufferSize <= short.MaxValue; i++, bufferSize = 4096 * i)
             {
-                buffer = ArrayPool<char>.Shared.Rent(bufferSize);
-                uint pathLength;
-                fixed (char* lpFilename = buffer)
+                lbuffer = ArrayPool<char>.Shared.Rent(bufferSize);
+                fixed (char* lpFilename = lbuffer)
                 {
-                    pathLength = GetModuleFileName(hModule, lpFilename, (uint)buffer.Length);
+                    pathLength = GetModuleFileName(hModule, lpFilename, (uint)lbuffer.Length);
+                    isBufferTooSmall = (WIN32_ERROR)Marshal.GetLastWin32Error() == WIN32_ERROR.ERROR_INSUFFICIENT_BUFFER;
                 }
 
                 if (pathLength == 0)
                 {
-                    ArrayPool<char>.Shared.Return(buffer);
+                    ArrayPool<char>.Shared.Return(lbuffer);
                     return string.Empty;
                 }
 
                 // If the length equals the buffer size we need to check to see if we were told the buffer was insufficient (it was trimmed)
-                if (pathLength < buffer.Length - 1 || (WIN32_ERROR)Marshal.GetLastWin32Error() != WIN32_ERROR.ERROR_INSUFFICIENT_BUFFER)
+                if (pathLength < lbuffer.Length - 1 && !isBufferTooSmall)
```

Comment:
```suggestion
                if (pathLength < lbuffer.Length)
```

I don't know why you had the `- 1` as part of the expression. That looks like a bug.